pkg/controllers/localvolumeset/localvolumeset_controller.go

```diff
@@ -127,17 +137,38 @@ func (r *LocalVolumeSetReconciler) syncStorageClass(ctx context.Context, lvs *lo
 			Labels: map[string]string{
 				common.OwnerNameLabel:      lvs.GetName(),
 				common.OwnerNamespaceLabel: lvs.GetNamespace(),
+				common.OwnerKindLabel:      localv1alpha1.LocalVolumeSetKind,
 			},
 		},
 		Provisioner:       "kubernetes.io/no-provisioner",
 		ReclaimPolicy:     &deleteReclaimPolicy,
 		VolumeBindingMode: &firstConsumerBinding,
 	}
+	currentStorageClass := &storagev1.StorageClass{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: lvs.Spec.StorageClassName}, currentStorageClass); err != nil {
+		if kerrors.IsNotFound(err) {
+			err := r.Client.Create(ctx, storageClass)
+			if err != nil && !kerrors.IsAlreadyExists(err) {
+				return err
+			}
+		} else {
+			return fmt.Errorf("failed to get storageclass %q: %v", lvs.Spec.StorageClassName, err)
+		}
+	}
 
-	err := r.Client.Create(ctx, storageClass)
-	if err != nil && !kerrors.IsAlreadyExists(err) {
-		return err
+	ownerName := currentStorageClass.Labels[common.OwnerNameLabel]
+	ownerNamespace := currentStorageClass.Labels[common.OwnerNamespaceLabel]
+	_, hasOwnerKind := currentStorageClass.Labels[common.OwnerKindLabel]
+
+	// Only update the missed owner kind label storageclass which is used for upgrade scenario
+	if ownerName == lvs.GetName() &&
+		ownerNamespace == lvs.GetNamespace() &&
+		!hasOwnerKind {
+		if err := r.Client.Update(ctx, storageClass); err != nil {
+			return fmt.Errorf("failed to update storageclass %q: %v", lvs.Spec.StorageClassName, err)
+		}
```

Comment:
I don't think this code will update the labels after upgrade like you intended.
If an existing SC has the old `local.storage.openshift.io/localvolumeset-owner-name` label then `ownerName == lvs.GetName()` will return false and the labels will not be updated.
IMO, you could use the same `applyStorageClass` function that is used for LocalVolumes:
https://github.com/openshift/local-storage-operator/blob/fac0f724c984f6d1ca3948236438244c6ce36915/pkg/controllers/localvolume/resourceapply.go#L14-L20
I suggest moving that into `pkg/common` and then call it from both controllers. That would also address my other comment above.

Reply:
In your upgrade scenario test, `my-lvset` already has the labels `local.storage.openshift.io/owner-name=my-lvset,local.storage.openshift.io/owner-namespace=openshift-local-storage` at the start of the test.
But if it instead has `local.storage.openshift.io/localvolumeset-owner-name=my-lvset,local.storage.openshift.io/localvolumeset-owner-namespace=openshift-local-storage` at the start of the test then the labels won't be updated.

Reply:
Good point, I checked it seems the const `ownerNameLabel = "local.storage.openshift.io/owner-name"` never used even in 4.6 we always use `local.storage.openshift.io/owner-name`label for lvset owned storageclass https://github.com/openshift/local-storage-operator/blob/release-4.6/pkg/controller/localvolumeset/reconcile.go#L103-L104. So we only need to add kind owner label for old sc, but I agree with you using apply will be even better. I'm also considering use apply when I try to fix this while it needs add the apiUpdater interface so I do not use it to make less changes. Let me update with the same approach with lv syn storageclass.